src/macaron/repo_finder/provenance_finder.py

```diff
@@ -83,13 +84,19 @@ def find_provenance(
             logger.debug("No provenance discovery functions provided/found for %s", purl)
             return []
 
-        for discovery_function in discovery_functions:
-            if parameters:
-                provenance = discovery_function(*parameters)
-            else:
-                # Assume any discovery function requires at least the PURL.
-                # None is required to prevent pylint error (E1120: no-value-for-parameter).
-                provenance = discovery_function(purl, None)
+        for index, discovery_function in enumerate(discovery_functions):
+            parameter_list = parameter_lists[index] if parameter_lists and index < len(parameter_lists) else [purl]
+            function_signature = signature(discovery_function)
+            if len(function_signature.parameters) != len(parameter_list):
+                logger.debug(
+                    "Mismatch between function and parameters: %s vs. %s",
+                    len(function_signature.parameters),
+                    len(parameter_list),
+                )
+                continue
```

Comment:
Hmmm I can see that:
- Because we are passing the function pointers and the corresponding parameters separately, we need to do a lot of checking. 
- We also need to handle the default parameter when the parameter list is empty for each individual discovery function, which is obscured from the caller of `find_provenance`. 
```python
parameter_list = parameter_lists[index] if parameter_lists and index < len(parameter_lists) else [purl]
```

I think to address such problems, we can instead provide the function pointer and its corresponding parameters together :thinking: 

This is what I had in mind (logs are removed for clarity):
```python3
    def find_provenance(
        self,
        purl: PackageURL,
        discovery_functions: list[partial[list[InTotoPayload]]] | None = None,
    ) -> list[InTotoPayload]:
        if not discovery_functions:
            if determine_abstract_purl_type(purl) == AbstractPurlType.REPOSITORY:
                # Do not perform default discovery for repository type targets.
                return []
            if purl.type == "npm":
                if not self.npm_registry:
                    logger.debug("Missing npm registry to find provenance in.")
                    return []
                discovery_functions = [
                    partial(
                        find_npm_provenance,
                        *(purl, self.npm_registry),
                    ),
                ]
            elif purl.type in ["gradle", "maven"]:
                if not self.jfrog_registry:
                    logger.debug("Missing npm registry to find provenance in.")
                    return []

                discovery_functions = [
                    partial(
                        find_gav_provenance,
                        *(purl, self.jfrog_registry),
                    )
                ]
        if not discovery_functions:
            return []

        for discovery_function in discovery_functions:
            provenance = discovery_function()
            if provenance:
                return provenance

        logger.debug("No provenance found.")
        return []
```
In this example, I leveraged `partial` (https://docs.python.org/3/library/functools.html#functools.partial) which support the use case of defining the function pointer and its parameter lists for being called at a later stage. Doing it this way, we are essentially:
- Leave the responsibility of defining the function call to the caller of `find_provenance` . For example, this is how someone would use this function if they want to use specific discovery functions instead of the default one:
```python
provenances = ProvenanceFinder.find_provenance(
    purl,
    [
        partial(discovery1, a, b, c),
        partial(discovery2, a, e, f),
    ],
)
```
- `partial` supports both positional and keyword arguments, which I think can allow us flexibly define the discovery function interface.  
- We can still have the default usage as `provenances = ProvenanceFinder.find_provenance(purl)`
- In the example, I also removed setting the default parameter list to `[purl]` if the parameter list provided is empty. The caller of `find_provenance` know the signature of the discovery functions, so they should take responsibility to provide the necessary parameters (from what I see, `find_provenance` don't create or obtain any extra parameters required for the discover functions, so I guess it is fine). Please let me know what you think about this idea.

Furthermore, I think we can separate `find_provenance` as I described above into 2 methods:
- `find_provenance(purl: PackageURL)` - the default find provenance logic
- `find_provenance_custom(purl: PackageURL, discovery_functions: list[partial[list[InTotoPayload]]]` - the more advanced method to iterate through `discovery_functions` and return the first list of provenances found.
However, I don't have a strong opinion on this change.

Reply:
@behnazh-w Please let me know what you think about it too, and whether it's inline with your initial design.

Reply:
I have addressed this mostly as suggested. The key difference is the `find_provenance_custom` equivalent being an internal function, as having it being public made little sense when there was no mechanism to supply the custom discovery functions. Instead, the `find_provenance` function will again be responsible for calling the appropriate function(s), and can be extended for other types as required moving forwards.